src/steamnetworkingsockets/clientlib/steamnetworkingsockets_p2p_ice.cpp

```diff
@@ -177,10 +177,79 @@ void CConnectionTransportP2PICE::Init()
 	cfg.m_pStunServers = vecStunServersPsz.data();
 
 	// Get the TURN server list
+	std_vector<std::string> vecTurnServers;
+	std_vector<std::string> vecTurnUsers;
+	std_vector<std::string> vecTurnPasses;
+	std_vector<ICESessionConfig::TurnServer*> vecTurnServersPsz;
 	if ( P2P_Transport_ICE_Enable & k_nSteamNetworkingConfig_P2P_Transport_ICE_Enable_Relay )
 	{
 		// FIXME
 		//cfg.m_nCandidateTypes = m_nAllowedCandidateTypes;
+
+		m_nAllowedCandidateTypes |= k_EICECandidate_Any_HostPublic | k_EICECandidate_Any_Reflexive;
+
+		{
+			CUtlVectorAutoPurge<char*> tempTurnServers;
+			V_AllocAndSplitString(m_connection.m_connectionConfig.m_P2P_TURN_ServerList.Get().c_str(), ",", tempTurnServers);
+			for (const char* pszAddress : tempTurnServers)
+			{
+				std::string server;
+
+				// Add prefix, unless they already supplied it
+				if (V_strnicmp(pszAddress, "turn:", 5) != 0)
+					server = "turn:";
+				server.append(pszAddress);
+
+				vecTurnServers.push_back(std::move(server));
+			}
+
+			// populate usernames
+			CUtlVectorAutoPurge<char*> tempTurnUsers;
+			V_AllocAndSplitString(m_connection.m_connectionConfig.m_P2P_TURN_UserList.Get().c_str(), ",", tempTurnUsers);
+			for (const char* user : tempTurnServers)
+			{
+				std::string server;				
+				server.append(user);
+				vecTurnUsers.push_back(std::move(server));
+			}
+
+			// populate passwords
+			CUtlVectorAutoPurge<char*> tempTurnPasses;
+			V_AllocAndSplitString(m_connection.m_connectionConfig.m_P2P_TURN_PassList.Get().c_str(), ",", tempTurnPasses);
+			for (const char* pass : tempTurnPasses)
+			{
+				std::string server;
+				server.append(pass);
+				vecTurnPasses.push_back(std::move(server));
+			}
+		}
+		if (vecTurnServers.empty())
+			SpewWarningGroup(LogLevel_P2PRendezvous(), "[%s] Reflexive candidates enabled by P2P_Transport_ICE_Enable, but P2P_STUN_ServerList is empty\n", ConnectionDescription());
+		else
+			SpewVerboseGroup(LogLevel_P2PRendezvous(), "[%s] Using STUN server list: %s\n", ConnectionDescription(), m_connection.m_connectionConfig.m_P2P_TURN_ServerList.Get().c_str());
+
+		Assert(vecTurnServers.size() == vecTurnUsers.size() == vecTurnPasses.size());
```

Comment:
Note: This doesn't actually work. :)  Probably need to make sure that a misconfiguration cannot lead to an access violation or leak of any credentials.  Probably just ignore all TURN servers  of the three arrays don't have the same length, rather than asserting.  Or...maybe if the users and passes list is empty, then that's OK, just treat them all as unauthenticated.

Or is there a standard syntax used by applications for specifying the configuration of a TURN server in a single string?    Like username@address ?

Reply:
I think if users and passes list is empty, treating them as unathenticated makes sense, (we set them as empty strings because of the below check.) And even if the three arrays don't have the same length, the code may take TURN servers array as a base, and populate or reduce size of related users and passes arrays accordingly to be sure all three having the same size?
This approach of course not covering the if password has a comma in it. 

May be best would be the ConfigValue struct supports non-primitive type which to have an array struct type of a TurnConfig for ConnectionConfig struct, force users to initialize configs seperately as an array of TurnConfig struct, and after init, the reader (CConnectionTransportP2PICE::Init method) will read the config as an array of TurnConfig, so using comma would work too?

I will think about it and look around for a standard usage if exists.

https://github.com/ValveSoftware/GameNetworkingSockets/blob/f98cd32f62877d31a3fe27fc91578e13f4f795c9/src/external/steamwebrtc/ice_session.cpp#L292

Reply:
Probably best to keep it simple strings, struct is complicated to maintain.  If there is an existing standard, then a single var with a string for each server sounds nice.  Otherwise, separate strings with the rule that either they need to be the same length, or the user+pass arrays both need to empty, sounds good.

Might need to make sure that V_AllocAndSplitString will do the right thing if one item is empty, e.g. "username1,,username3"

Reply:
Okay i ll make sure V_AllocAndSplitString will cover the cases.

Yes, if there is an existing standard, having a single string per server would be cool. From my fast research i see JSON kinda formats exists as below; '?' is used to seperate transport and turn maybe, user+pass are in different places. But i will look more.

Okay, makes sense, if array lengths are not same, then it means user+pass arrays need to empty. If lengths are equal, then is already okay.

```
 customServers: [
      {
        urls: [
          'turn:123.124.125.126:3478?transport=udp’',
          'turn:123.124.125.126:3478?transport=tcp'
        ],
        username: 'webrtc',
        credential: 'foO0Bar1'
      },

List<IceServer> serverList = new IceServer(
  'turn:123.124.125.126:3478?transport=udp', // TURN server URL
  'webrtc', // Username
  'foO0Bar1' // Credential
);

```


I will come up with

Reply:
I tested the V_AllocAndSplitString method and it handles it nicely.
ex: "username1,,username3" populates array of 2, values are verified by debugger.
ex: "asd,asdasd,,sd,,sdsd" populates array of  4. values are verified by debugger.
ex: "asd,asdasd,,,,,sd,,sdsd" populates array of  4. values are verified by debugger.

If user+pass arrays length are not same with turnservers array length, then it treat all the TURN servers as unauthenticated.

I pushed the commits. Thank you!

Reply:
I think we will need to pass bIncludeEmptyStrings=true to V_AllocAndSplitString.  Otherwise, there's no way to have some TURN servers that require a username and password, and other servers that don't.

Reply:
Yes, you are right. Lemme cover this. This was on my mind at the morning btw. thumbs up.

Reply:
Enabling bIncludeEmptyStrings generates 1 string from "username1," and generates 2 strings from "username1,,"
We need 2 strings "from username1," and 3 strings from "username1,,"

I added the below _else if block_ at V_SPlitStrings2 which seperates "username1," into 2 strings instead of 1. Result is "username1" and "". It works.

https://github.com/LeventSevgili/GameNetworkingSockets/blob/b5ce6785abe11b9071e8570d20ae5515240288a7/src/vstdlib/strtools.cpp#L405

I tested below cases:

"username1,,username3" -> "username1" "" "username2" 3 strings.
"pass1,,pass3" -> "pass1" "" "pass3" as 3 strings.
OR
"username1,username2," -> "username1" "username2" "" 3 strings.
"pass1,pass2," -> "pass1" "pass2" "" 3 strings.

What do you think? Thanks.